Add new tooling

tumidi opened the pull request:
- remove tox (lives in [qpy-dev](https://github.com/questionpy-org/qpy-dev) now)
- add ruff as replacement for pylint/flake8
- reformat codebase
- fix linter errors
- support for Python ~3.9, 3.10~ 3.11, **3.12**
- comply to questionpy-org/questionpy-docs/pull/5
- use update GitHub workflow

tumidi committed:
test: fix paths in tests

Ensures tests pass regardless of cwd.

tumidi committed:
chore(tooling): add ruff, remove pylint and flake8

Split up dev dependencies into separate groups, so they can be
installed separately.

tumidi committed:
chore(tooling): add .editorconfig file

tumidi committed:
chore(tooling): add ruff_defaults.toml

tumidi committed:
chore(tooling): remove tox

tumidi committed:
chore: remove left-over coverage.sh

tumidi committed:
chore: ruff autofixes (only safe)

tumidi committed:
chore: fix linter errors (F, E, W, PL)

tumidi committed:
chore: fix linter errors (more trivial ones)

tumidi committed:
refactor: split up modules to avoid circular import

Put `ensure_package_and_question_state_exist` into separate
module`decorators.py`.

tumidi committed:
chore: fix linter errors (UP)

tumidi committed:
chore: fix linter errors (TCH)

tumidi committed:
chore: fix remaining linter errors

tumidi committed:
chore: fix linter errors (S101: turn asserts into exceptions)

tumidi committed:
chore: set GitHub actions workflow to `v8.1` branch

tumidi committed:
docs: update to new questionpy_docs

tumidi committed:
build: support Python versions 3.11 and 3.12

tumidi committed:
chore: reformat using ruff

MHajoha commented on questionpy_server/worker/runtime/manager.py:
```diff
@@ -99,7 +99,10 @@ def loop(self) -> None:
 
     def on_msg_load_qpy_package(self, msg: LoadQPyPackage) -> MessageToServer:
         self._require_init(msg)
-        assert self.worker_type
+
+        if not self.worker_type:
+            errmsg = "'worker_type' is not set."
+            raise RuntimeError(errmsg)
```
Das ist eigentlich genau der korrekte Use-Case für Assertions: `_require_init` stellt sicher, dass `self._worker_type` gesetzt ist, aber das versteht der Typechecker nicht. (Deswegen heißt es auch Assertion. Du prüfst nicht, ob das so ist, sondern behauptest es)

Als Kompromiss wäre es vllt. möglich, `if not self._worker_type: self._raise_not_initialized(msg)` oder so zu schreiben. Ich habe es nicht ausprobiert, aber mit `-> NoReturn` dürfte Mypy das dann ja verstehen. Am liebsten aber die Regel ausmachen.

tumidi replied:
Es gibt an verschiedenen Stellen Refaktoriersungsbedarf. Ich hab das aber bewusst mal ausgeklammert, um die Commits möglichst atomar zu halten.

> Du prüfst nicht, ob das so ist, sondern behauptest es

Aber ja nur konzeptionell. Während der Laufzeit ist es ein Runtime-Check, also syntaktischer Zucker für:
```
if __debug__:
    if not self.worker_type: raise AssertionError
```

`assert` ist an der Stelle "okay-ish": Führt halt zu einem unnötigen redundanten Runtime-Check. Der Check passiert ja bereits in `_require_init`. Ich denke, besser (ohne Laufzeit-Auswirkung) wäre:

```
worker_type = cast(str, self.worker_type)
```

Aber auch das wäre nicht DRY, da es ja in jedem message handler passiert, was zeigt, dass es so eh nicht ganz sauber gelöst ist. Besser wäre es, die Stellen so zu refaktorisieren, dass das repetitive type narrowing unnötig wird.

1. [getter zu verwenden](https://github.com/questionpy-org/questionpy-server/commit/98ac46b40a314671961f28558b47467fea6610db)
   Würde allerdings die Möglichkeit verlieren die empfangene Message in der Exception zu referenzieren.
2. [runtime check methoden in decorators umwandeln](https://github.com/questionpy-org/questionpy-server/commit/ecfbc2c068e872d04bac7e5dc515af91f797850d)
3. gibt natürlich noch andere Möglichkeiten, zB. den state vom `WorkerManager` zu trennen und wegzuabstrahieren, bspw.
   uninitialisierter State → (bootstrap) → initialisierter State → (on_msg_load_qpy_package) → State mit main package, usw.
   Das wäre evtl. eh expliziter, anstatt zwischendrin zu checken welches Attribut wo genau jetzt `None` ist
   und dadurch implizit abzuleiten in welchem State der Manager gerade ist...

tumidi replied:
Davon abgesehen steht halt die Frage im Raum, ob wir `assert` in Production-Code zulassen, also S101 deaktivieren wollen. `assert`s verschwinden ja, sobald man Python byte code mit `-O` erstellt, daher ist das auch eine [Security-Regel aus Bandit](https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html).
Ob man die jetzt braucht, da kann man geteilter Ansicht sein. Bin da nicht festgelegt. Persönlich würde ich einfach auf `assert` in Production-Code verzichten, da ich keinen Mehrwert sehe, insbesondere wenn man mit MyPy weitgehend type safety bekommt... :shrug:

MHajoha replied:
> Ich denke, besser (ohne Laufzeit-Auswirkung) wäre:
> 
> ```
> worker_type = cast(str, self.worker_type)
> ```

Casts finde ich für sowas etwas gefährlich, weil so ein Schusselfehler ohne Beanstandung durch geht:
```python
question_type = cast(BaseQuestionType, self.worker_type)
```

Mypy prüft leider nicht, ob `str | None` und `BaseQuestionType` irgendeine Schnittmenge haben

> Aber auch das wäre nicht DRY, da es ja in jedem message handler passiert, was zeigt, dass es so eh nicht ganz sauber gelöst ist. Besser wäre es, die Stellen so zu refaktorisieren, dass das repetitive type narrowing unnötig wird.
> 
>     1. [getter zu verwenden](https://github.com/questionpy-org/questionpy-server/commit/98ac46b40a314671961f28558b47467fea6610db)
>        Würde allerdings die Möglichkeit verlieren die empfangene Message in der Exception zu referenzieren.

Wäre schon schön, wenn wir die beibehalten könnten

>     2. [runtime check methoden in decorators umwandeln](https://github.com/questionpy-org/questionpy-server/commit/ecfbc2c068e872d04bac7e5dc515af91f797850d)

Nette Idee, finde ich zumindest für den Moment aber noch etwas schwergewichtig.

>     3. gibt natürlich noch andere Möglichkeiten, zB. den state vom `WorkerManager` zu trennen und wegzuabstrahieren, bspw.
>        uninitialisierter State → (bootstrap) → initialisierter State → (on_msg_load_qpy_package) → State mit main package, usw.
>        Das wäre evtl. eh expliziter, anstatt zwischendrin zu checken welches Attribut wo genau jetzt `None` ist
>        und dadurch implizit abzuleiten in welchem State der Manager gerade ist...

Das gefällt mir, ist vermutlich der sauberste Ansatz.

Ich würde vorschlagen, erstmal entweder Assertions beizubehalten oder `if not self.worker_type: self._raise_not_initialized(msg)` zu machen, je nachdem wie stark deine Meinung dazu ist, und ein größeres Refactoring a la 3. für die Zukunft im Hinterkopf zu behalten. Mit casts könnte ich mich auch noch anfreunden, aber s.o. sorgt das hier halt paradoxerweise für _weniger_ type safety.

tumidi replied:
> Casts finde ich für sowas etwas gefährlich

Ja, richtig. Cast sollte auch kein Vorschlag sein. Habe ich mich schlecht ausgedrückt.

Der Vorschlag von dir ist doch super: `if not self.worker_type: self._raise_not_initialized(msg)`
Verzichtet auf `assert`, braucht kein doppelten Runtime check und bedarf keiner großen Änderung. 👍

MHajoha replied:
Knaller, danke dir :+1:

MHajoha commented on questionpy_server/decorators.py:
```diff
@@ -39,10 +39,9 @@ def decorator(function: RouteHandler) -> RouteHandler:
-            assert issubclass(
-                main_part_json_model, MainBaseModel
-            ), f"Parameter 'data' of function {function.__name__} has unexpected type."
+        if main_part_json_model and not issubclass(main_part_json_model, MainBaseModel):
+            msg = f"Parameter 'data' of function {function.__name__} has unexpected type."
+            raise TypeError(msg)
```
wrt <https://docs.astral.sh/ruff/rules/f-string-in-exception/>

Hier den Code weniger leserlich zu machen, damit der Stacktrace minimal klarer ist, sehe ich nicht ganz. Ist keine starke Meinung, aber ich würde die Regel lieber global ausmachen.

tumidi replied:
Wird der Code weniger leserlich? Finde entschachtelte Instruktionen tendenziell eher förderlich für die Lesbarkeit, z.B.

```py
msg = f"'{InitWorker.__name__}' message expected, '{type(init_msg).__name__}' received"
raise WorkerNotInitializedError(msg)

# anstatt

raise WorkerNotInitializedError(
    f"'{InitWorker.__name__}' message expected, '{type(init_msg).__name__}' received"
)
```

Aber ich hab da auch keine starke Meinung zu. Mein Ansatz war soweit sinnvoll alle Regeln auf default/recommended zu belassen und von da aus auszudünnen.

MHajoha replied:
Ich habe nicht daran gedacht, dass mit Variante 2 an manchen Stellen mehr Umbrüche notwendig sind. Dann ziehe ich das zurück, nimmt sich glaube ich wirklich nicht viel. Den Ansatz finde ich übrigens korrekt, das hier war keine Kritik (bzw nur an der Regel, nicht an dir), sondern eben ein Vorschlag, was ausgedünnt werden könnte ^^

MHajoha commented on ruff_defaults.toml:
```diff
@@ -0,0 +1,132 @@
+
+[lint.per-file-ignores]
+"**/scripts/*" = ["INP001", "T201"]
+"**/tests/**/*" = ["PLC1901", "PLR2004", "S", "TID252", "FBT"]
```
Bombe, warum kann pylint das nicht

MHajoha reviewed (approved):
Insgesamt finde ich ruff ziemlich geil :+1:

MHajoha commented:
Magst du gleich noch watchdog auf `^4.0.1` hochziehen? Tests laufen zwar durch, ["offiziell" unterstützt watchdog Python 3.12 aber erst seit 4.0.](https://github.com/gorakhargosh/watchdog/blob/master/changelog.rst#400). Und es kollidiert sonst mit qpy-dev, da steht schon 4.0.0 drin.

tumidi committed:
chore: fix separated f-strings (left-over artifacts from reformatting)

tumidi committed:
chore: update to watchdog 4.0.0

Update implementation of `directory_iterator` to account for
signature change of `watchdog.utils.dirsnapshot.DirectorySnapshot`.

tumidi committed:
refactor(WorkerManager): better handling of uninitialized attributes

Avoid checking for falsy value twice in message handlers.

tumidi committed:
chore: update questionpy-common
